[Siew Hui Zhuan] iP

huizhuansam opened the pull request:
# Pull Request 
This is my pull request. Look at it! So many commits! It must be an *amazing* pull request then. 

## What's Good 👍🏼 
- Search tasks by keyword
- Search tasks by date
- Save tasks to a file so you can quit the program, start it up and **boom** your tasks are still there (\**click*\* **nice**)

## Things to Note 📓 
1. This program might be buggy
2. I absolutely loathed writing the JUnit tests
3. **2.** will be the reason for **1.** 

## Impressive Stuff 🥵 
`TaskList.java` line 118 - 132, where I attempted to write a functional flavored class method:
```java
/**
 * Returns a TaskList containing Task that satisfy the supplied predicate.
 *
 * @param predicate The predicate to test the Task with.
 * @return A TaskList containing Task that satisfies the given predicate.
 */
public TaskList filter(Function<Task, Boolean> predicate) {
    ArrayList<Task> res = new ArrayList<>();
    for (int i = 0; i < size(); i++) {
        if (predicate.apply(list.get(i))) {
            res.add(list.get(i));
        }
    }
    return new TaskList(res);
}
```
so that I can write the find task method in `FindCommand.java` as such:
```java
/**
 * Prints the tasks matching the keyword.
 *
 * @param tasks The list of tasks.
 * @param ui The UI object.
 * @param storage The storage utility.
 */
@Override
public void execute(TaskList tasks, Ui ui, Storage storage) {
    ui.print("Here are the tasks containing \"" + keyword + "\":");
    ui.print(tasks.filter((task) -> task.keywordMatcher(keyword)).toStringArray());
}
```
Functional programming is so fun! ~~Java isn't, though~~.

## Todo ✅ 
- [x] Write this PR description
- [x] Finish week 4 IP

## My Favorite Computing Quote 🗣️ 
> I know Gilfoyle probably came in here and puked out a bunch of tech specs, three-fourths of which are total horses***. Did he bring up the Iranian revolution thing? Yeah, those words mean nothing. But here's a fact, I'm the only one of these clowns that can code in Java and I write sleek performant low-overhead Scala code with higher-order functions that will run on anything. Period. End of sentence.
> -- <cite>Dinesh Chugtai, HBO Silicon Valley Season 1 Episode 2 (The Cap Table)</cite>

## Apology Video (Gone Wrong) (Gone Sussy) 🙇🏼  😢  
[Link to my apology video](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Commit:
Add Gradle support

huizhuansam committed:
Level 1 (#1)

* Update .gitignore to not track Java class files

* Add Printer class

* Modify divider line

* Add basic IO logic

huizhuansam committed:
Level 2 (#2)

* Add memory class

* Add processor class

* Modify Duke to direct input to processor

huizhuansam committed:
A classes (#3)

* Add task class

* Modify memory to hold array of tasks

huizhuansam committed:
Add ability to mark task as done (#4)

huizhuansam committed:
Refactor to support inheritance (#5)

huizhuansam committed:
Level 4 (#6)

* Add input parser

* Add factory method

* Add feedback printer

* Modify add task method to accept task subclass

huizhuansam committed:
Modify test files (#7)

huizhuansam committed:
Add exception classes (#8)

huizhuansam committed:
Add error handling (#9)

huizhuansam committed:
Refactor memory to use ArrayList (#10)

huizhuansam committed:
Add delete task functionality (#11)

huizhuansam committed:
Add enum types for commands (#12)

huizhuansam committed:
Update test files (#13)

huizhuansam committed:
Fix method access modifiers (#14)

huizhuansam committed:
Extract input parser to new class

huizhuansam committed:
Refactor exception handling

huizhuansam committed:
Enforce enum as parser parameter

huizhuansam committed:
Add javadocs

huizhuansam committed:
Add toString method

huizhuansam committed:
Add javadocs

huizhuansam committed:
Refactor exception handling

huizhuansam committed:
Add UI class

huizhuansam committed:
Add Tasklist class

huizhuansam committed:
Add command class

huizhuansam committed:
Add datedtask class

huizhuansam committed:
Add enum types for command and tasks

huizhuansam committed:
Refactor event class to subclass datedtask

huizhuansam committed:
Add parser methods

huizhuansam committed:
Add oop style to Duke

huizhuansam committed:
Deprecate Memory, Printer, Processor class

huizhuansam committed:
Update test files

huizhuansam committed:
Merge pull request #24 from huizhuansam/A-MoreOOP

A-MoreOOP

huizhuansam committed:
Refactor java files to packages

huizhuansam committed:
Rename packages to adhere to convention

huizhuansam committed:
Update test file

huizhuansam committed:
Merge pull request #25 from huizhuansam/A-Packages

A-Packages

huizhuansam committed:
Update .gitignore to ignore program data

huizhuansam committed:
Add file not found exception

huizhuansam committed:
Add storage class

huizhuansam committed:
Add storage as method parameters

huizhuansam committed:
Add date class

huizhuansam committed:
Add invalid date exception

huizhuansam committed:
Refactor tasks to use date object

huizhuansam committed:
Fix delete index bug

huizhuansam committed:
Add date equals method

huizhuansam committed:
Add filter task by date function

huizhuansam committed:
Update .gitignore to ignore test data

huizhuansam committed:
Add JSON-simple library

huizhuansam committed:
Add feature to parse tasks to JSON file

huizhuansam committed:
Fix remove index off by one error

huizhuansam committed:
Add UI response to successful load

huizhuansam committed:
Add feature to read from JSON file

huizhuansam committed:
Fix parse to todo isdone bug

huizhuansam committed:
Merge pull request #27 from huizhuansam/branch-Level-7

Level-7

huizhuansam committed:
Merge branch 'master' into branch-Level-8

huizhuansam committed:
Merge pull request #28 from huizhuansam/branch-Level-8

Level-8

huizhuansam committed:
Fix breaking changes from Level 7 and Level 8 (#29)

* Fix JSON encoding bug

* Fix additional error handling

* Fix data type from string to date

* Update test file

huizhuansam committed:
A-JUnit (#30)

* Add test for Todo class

* Add test for event class

* Add test for deadline class

* Add test for date class

* Add test for AddCommand

huizhuansam committed:
Add gradle build main class (#31)

huizhuansam committed:
Add Javadocs to Duke file

huizhuansam committed:
Add Javadocs to utility classes

huizhuansam committed:
Add Javadocs to Task classes

huizhuansam committed:
Add Javadocs to Date class

huizhuansam committed:
Add Javadocs to Commands class

huizhuansam committed:
Merge pull request #32 from huizhuansam/branch-A-JavaDoc

Update Javadocs

huizhuansam committed:
Remove unused import

huizhuansam committed:
Merge pull request #33 from huizhuansam/branch-A-CodingStandard

Adhere to Java coding standard

huizhuansam committed:
Add find tasks functionality

huizhuansam committed:
Merge branch 'master' into branch-Level-9

huizhuansam committed:
Merge pull request #34 from huizhuansam/branch-Level-9

Add find tasks functionality

huizhuansam committed:
Fix missing Javadocs

huizhuansam committed:
Merge pull request #35 from huizhuansam/fix-javadocs

Fix missing Javadocs

huizhuansam committed:
Add Checkstyle configuration files

huizhuansam committed:
Fix code style violation at Storage.java

huizhuansam committed:
Fix code style violation at Duke.java

huizhuansam committed:
Fix import order violation at Command.java

huizhuansam committed:
Fix import order violation at ExitCommand.java

huizhuansam committed:
Fix trailing whitespace at ListCommand.java

huizhuansam committed:
Fix import order and method name convention violation

huizhuansam committed:
Fix attribute naming violation in Commands.java

huizhuansam committed:
Fix attribute naming violation in Tasks.java

huizhuansam committed:
Fix import order violation at Deadline.java

huizhuansam committed:
Fix import order violation at Event.java

huizhuansam committed:
Fix import order violation at TaskList.java

huizhuansam committed:
Fix method name violation

huizhuansam committed:
Fix import order violation at AddCommandTest.java

huizhuansam committed:
Fix import order violation at DateTest.java

huizhuansam committed:
Fix import order violation at TaskStub.java

huizhuansam committed:
Fix import order violation at DeadlineTest.java

huizhuansam committed:
Fix import order violation at EventTest.java

huizhuansam committed:
Fix import order violation at TodoTest.java

huizhuansam committed:
Merge pull request #40 from huizhuansam/A-CheckStyle

A-CheckStyle

huizhuansam committed:
Deprecate text ui test files

huizhuansam committed:
Update README.md to include acknowledgements

huizhuansam committed:
Move Duke.java to package

huizhuansam committed:
Update README.md

huizhuansam committed:
Add JavaFX Gradle dependencies

huizhuansam committed:
Add user and bot avatars

huizhuansam committed:
Add fxml files

huizhuansam committed:
Add launcher and main class

huizhuansam committed:
Add dialog box and main window class

huizhuansam committed:
Setup GUI

huizhuansam committed:
Deprecate text UI class

huizhuansam committed:
Add dialog styling

huizhuansam committed:
Deprecate text UI test stub

huizhuansam committed:
Update README.md

huizhuansam committed:
Merge pull request #41 from huizhuansam/branch-Level-10

Level-10

huizhuansam committed:
Fix missing Javadocs (#42)

huizhuansam committed:
Fix method names

huizhuansam committed:
Fix Ainsley last name

huizhuansam committed:
Fix Ainsley's last name (again)

bharathcs commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+import duke.util.TaskList;
+
+/** An abstract class representing the user's commands. */
+public abstract class Command {
```
I like the way you've structured the Command class, the pure functional form is great.

bharathcs commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,60 @@
+/** Represents the "delete" command. */
+public class DeleteCommand extends Command {
+    /** The index of the Task to be deleted. */
+    private int index;
```
Just to make things clearer (for devs), maybe this should be final? Not that it really matters, pretty sure the compiler will optimise for it since its effectively final.

bharathcs replied:
I've also noticed this is other Command files (indices and dates), so do remember to check those out if you want to move on this suggestion.

huizhuansam replied:
You are right about this, but I think it is not terribly important to fix this now. Probably will do codebase optimization towards the end of the project, I'll take this into account. Thanks for the heads up.

bharathcs commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    public abstract String execute(TaskList tasks, Storage storage) throws IOException, NoSuchTaskException;
+
+    /** Returns true if the Command is an exit command, else false.*/
+    public abstract boolean isExit();
```
Makes sense, I like this!

bharathcs commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,116 @@
+import duke.exception.InvalidDateException;
+
+/** Represents an Event object. */
+public class Event extends DatedTask {
```
Hey, I think Events are supposed to have two dates (a to and from), and thats what differentiates them from deadline.

See [level 4 (Todos, Events, Deadlines)](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/projectDuke/index.html#level-4-todos-events-deadlines)

huizhuansam replied:
Nice catch, I somehow missed this part.

bharathcs commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /** Marks the task's status as done */
+    public Task markTaskAsDone() {
```
Just thinking, would it be neater if it was just `taskObj.setDone()` instead of `taskObj.markTaskAsDone()`?

huizhuansam replied:
In my opinion, `markTaskAsDone()` sounds more natural than `setDone()`, as it is more akin to real-life task lists that you will "mark" tasks as done. But I understand `setDone()` sounds more OO because getters and setters or whatnot. Will consider changing this for the sake of other devs looking at my code. 👍🏼

bharathcs commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,96 @@
+     * @return True if the description matches the keyword, false otherwise.
+     */
+    public final boolean matchKeyword(String keyword) {
+        return Pattern.compile(".*\\b" + keyword + "\\b.*").matcher(description).find();
```
Nicely done

bharathcs commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,89 @@
+import duke.exception.UnknownCommandException;
+
+/** Utility class that handles parsing of user input to program command. */
+public class Parser {
```
I really like how you did this, neat implementation.

bharathcs commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     *
+     * @param tasks The TaskList created from the instance of the Duke program.
+     */
+    @SuppressWarnings("unchecked")
```
Perhaps you should clarify in your javadocs on the reason for your warning suppression? (Note: I've seen this in a few other places too)

huizhuansam replied:
Good call. The warning suppression is because I can guarantee the correctness of the object type during runtime (line 73 of this file), but maybe other devs might get confused. I'll leave an explanatory comment here.

bharathcs commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     * @return An array of string representation of the Tasks in the list, numbered from 1.
+     */
+    public String[] toStringArray() {
+        String[] copy = list.stream().map(Object::toString).toArray(String[]::new);
+        for (int i = 0; i < size(); i++) {
+            copy[i] = (i + 1) + "." + copy[i];
+        }
```
Nice! Unfortunate we don't have combine for Java streams or you could use stream operations for the indexing too.

huizhuansam replied:
My feelings exactly... the code would've looked neater without the for loop, but it is Java, and with it comes clumsy syntax :(

bharathcs reviewed (commented):
LGTM! You've done a great job, and I love the functional style of your code.

> "Nice and spicy"
> -- Siew Hui Zhuan

huizhuansam commented:
@bharathcs Thanks for the review! You really pored through my code and picked out the sketchiest bits haha. Good eye for detail!

Neha-5678 commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @param tasks The list of tasks in the program.
+     * @param storage The storage utility.
+     * @throws NoSuchTaskException If the task does not exist at the specified index.
```
Maybe you could consider adding a description of what the method returns under Javadocs? I think you may have missed it out! You could consider making this change to all your command classes (classes that inherit from Command class) :)

huizhuansam replied:
Ah, I accidentally left this out after implementing my GUI. Thanks for pointing it out :)

Neha-5678 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,69 @@
+package duke.command;
+
+import duke.date.Date;
```
Maybe you could consider standardising the order of the import statements across your files to make it easier to browse through? Your code is really clean so I am picking on the minute details.

huizhuansam replied:
Can I clarify what standard are you referring to? IntelliJ and CheckStyle seem to disagree, so I went with CheckStyle's import rules (if you are wondering).

Neha-5678 commented on src/main/java/duke/enums/Tasks.java:
```diff
@@ -0,0 +1,34 @@
+package duke.enums;
```
You already did an amazing job for Javadocs and I am aware that the requirement was to add documentation to half of the files, however I just felt that it would be even better to add some documentation to this file if time persists so that it is easier to understand the code.

huizhuansam replied:
Noted, I will add more Javadocs if I have the time.

Neha-5678 commented on src/main/java/duke/exception/FileNotFoundException.java:
```diff
@@ -0,0 +1,4 @@
+package duke.exception;
```
Is this file necessary? Would just like to check!

huizhuansam replied:
Yep! This exception is used in `Storage.java`. I think I would probably add more explanatory comments on the exception classes I've created, so other devs know what these are for.

Neha-5678 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @param description The Deadline's description.
+     * @param by The date to complete the Task by.
+     * @param isDone The done status of the Task
```
I think you may have forgotten to add a punctuation behind this parameter description. Maybe you could add it in.

huizhuansam replied:
Nice catch.

Neha-5678 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Event constructor.
+     *
+     * @param description The event's description
```
Just a reminder to add punctuation for these parameter descriptions :)

huizhuansam replied:
👍🏼

Neha-5678 reviewed (commented):
You code is very neat and it was a very difficult task to spot coding violations. Good job! just a few nits to fix :-)

huizhuansam commented:
@Neha-5678 Thanks for the review!

huizhuansam committed:
Fix Pull Request Review (#43)

Minor bug fix and code cleanup.

Reviewers left some comments on the PR, the majority with regards to Javadocs and minor code style issues. One reviewer pointed out that I left out the end date of the Event task class.

Add missing Javadocs to every public and private method and explanatory comments on warning suppressions. Fix and refactor the program to support the Event task class to having start and end Date objects. Fix JUnit tests to accommodate this change.

My intention is to make the codebase more readable. This commit should resolve some of the comments I've received regarding my code style.

Changes made as commits:
* Add final to commands

* Fix missing end date property in Event

* Fix JUnit test files

* Add explanatory comments on suppressions

* Add missing return Javadocs

* Add Javadocs to enum classes

* Add missing punctuation mark

* Add missing punctuation marks

* Add Javadocs for exceptions

huizhuansam committed:
Fix code quality issues

huizhuansam committed:
Add assertion to Duke

huizhuansam committed:
Add assertion to tasks

huizhuansam committed:
Merge pull request #44 from huizhuansam/branch-Assertions

Add Assertions

huizhuansam committed:
Merge branch 'master' into branch-Code-Quality

huizhuansam committed:
Fix line length

huizhuansam committed:
Merge pull request #45 from huizhuansam/branch-Code-Quality

Code Quality

huizhuansam committed:
Add CI script (#46)

huizhuansam committed:
Add sort command

huizhuansam committed:
Implement sort routine

huizhuansam committed:
Merge pull request #47 from huizhuansam/C-Sort

Add Sort Task by Date

huizhuansam committed:
Fix gradle build

huizhuansam committed:
Update README.md

huizhuansam committed:
Add Ui.png

huizhuansam committed:
Add User Guide

huizhuansam committed:
Merge pull request #49 from huizhuansam/A-UserGuide

Add User Guide

huizhuansam committed:
Change Ui.png photo

huizhuansam committed:
Merge pull request #50 from huizhuansam/fix-update-ui-png

Change Ui.png photo
